{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "occupational-bruce",
   "metadata": {},
   "outputs": [],
   "source": [
    "from glob import glob\n",
    "import os\n",
    "import pandas as pd\n",
    "import json\n",
    "\n",
    "# table print options\n",
    "pd.set_option('display.max_rows', 100)\n",
    "pd.set_option('display.max_colwidth', 1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "wooden-multimedia",
   "metadata": {},
   "source": [
    "# Prepare ATAC-seq configuration file of .fastq data \n",
    "Parses data data directory names and .fastq files in proejct directory, extracting metadata.\n",
    "The script generates a per sample config .json file which is subsequently used for the ENCODE ATAC-seq pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "effective-enough",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = \"/lustre/scratch126/cellgen/team205/sk29/matthias_fb/data/atac_simon/proj068/FASTQ_Generation_2022-12-17_02_34_43Z-640535898\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "proud-artwork",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>path</th>\n",
       "      <th>basename</th>\n",
       "      <th>fastq_R1</th>\n",
       "      <th>fastq_R2</th>\n",
       "      <th>condition</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>lane</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>/lustre/scratch126/cellgen/team205/sk29/matthias_fb/data/atac_simon/proj068/FASTQ_Generation_2022-12-17_02_34_43Z-640535898/unconditioned_5_L001-ds.46edbd96addd4c7bbcf669fbdd201e29</td>\n",
       "      <td>unconditioned_5_L001-ds.46edbd96addd4c7bbcf669fbdd201e29</td>\n",
       "      <td>unconditioned-5_S5_L001_R1_001.fastq.gz</td>\n",
       "      <td>unconditioned-5_S5_L001_R2_001.fastq.gz</td>\n",
       "      <td>unconditioned</td>\n",
       "      <td>5</td>\n",
       "      <td>L001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>/lustre/scratch126/cellgen/team205/sk29/matthias_fb/data/atac_simon/proj068/FASTQ_Generation_2022-12-17_02_34_43Z-640535898/conditioned_DMSO_22_L002-ds.3f381fa67ef345fe81257f9d7fc3ed00</td>\n",
       "      <td>conditioned_DMSO_22_L002-ds.3f381fa67ef345fe81257f9d7fc3ed00</td>\n",
       "      <td>conditioned-DMSO-22_S16_L002_R1_001.fastq.gz</td>\n",
       "      <td>conditioned-DMSO-22_S16_L002_R2_001.fastq.gz</td>\n",
       "      <td>conditioned_DMSO</td>\n",
       "      <td>22</td>\n",
       "      <td>L002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>/lustre/scratch126/cellgen/team205/sk29/matthias_fb/data/atac_simon/proj068/FASTQ_Generation_2022-12-17_02_34_43Z-640535898/conditioned_DMSO_21_L002-ds.d6b480e7b7104216b998a896f63665dd</td>\n",
       "      <td>conditioned_DMSO_21_L002-ds.d6b480e7b7104216b998a896f63665dd</td>\n",
       "      <td>conditioned-DMSO-21_S15_L002_R1_001.fastq.gz</td>\n",
       "      <td>conditioned-DMSO-21_S15_L002_R2_001.fastq.gz</td>\n",
       "      <td>conditioned_DMSO</td>\n",
       "      <td>21</td>\n",
       "      <td>L002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>/lustre/scratch126/cellgen/team205/sk29/matthias_fb/data/atac_simon/proj068/FASTQ_Generation_2022-12-17_02_34_43Z-640535898/conditioned_Belinostat_18_L004-ds.898be0a8019544ce8e9206a811da1756</td>\n",
       "      <td>conditioned_Belinostat_18_L004-ds.898be0a8019544ce8e9206a811da1756</td>\n",
       "      <td>conditioned-Belinostat-18_S12_L004_R1_001.fastq.gz</td>\n",
       "      <td>conditioned-Belinostat-18_S12_L004_R2_001.fastq.gz</td>\n",
       "      <td>conditioned_Belinostat</td>\n",
       "      <td>18</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>/lustre/scratch126/cellgen/team205/sk29/matthias_fb/data/atac_simon/proj068/FASTQ_Generation_2022-12-17_02_34_43Z-640535898/conditioned_DMSO_19_L004-ds.f14158dcc3324931b3edfeb0245a862c</td>\n",
       "      <td>conditioned_DMSO_19_L004-ds.f14158dcc3324931b3edfeb0245a862c</td>\n",
       "      <td>conditioned-DMSO-19_S13_L004_R1_001.fastq.gz</td>\n",
       "      <td>conditioned-DMSO-19_S13_L004_R2_001.fastq.gz</td>\n",
       "      <td>conditioned_DMSO</td>\n",
       "      <td>19</td>\n",
       "      <td>L004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                                             path  \\\n",
       "0            /lustre/scratch126/cellgen/team205/sk29/matthias_fb/data/atac_simon/proj068/FASTQ_Generation_2022-12-17_02_34_43Z-640535898/unconditioned_5_L001-ds.46edbd96addd4c7bbcf669fbdd201e29   \n",
       "1        /lustre/scratch126/cellgen/team205/sk29/matthias_fb/data/atac_simon/proj068/FASTQ_Generation_2022-12-17_02_34_43Z-640535898/conditioned_DMSO_22_L002-ds.3f381fa67ef345fe81257f9d7fc3ed00   \n",
       "2        /lustre/scratch126/cellgen/team205/sk29/matthias_fb/data/atac_simon/proj068/FASTQ_Generation_2022-12-17_02_34_43Z-640535898/conditioned_DMSO_21_L002-ds.d6b480e7b7104216b998a896f63665dd   \n",
       "3  /lustre/scratch126/cellgen/team205/sk29/matthias_fb/data/atac_simon/proj068/FASTQ_Generation_2022-12-17_02_34_43Z-640535898/conditioned_Belinostat_18_L004-ds.898be0a8019544ce8e9206a811da1756   \n",
       "4        /lustre/scratch126/cellgen/team205/sk29/matthias_fb/data/atac_simon/proj068/FASTQ_Generation_2022-12-17_02_34_43Z-640535898/conditioned_DMSO_19_L004-ds.f14158dcc3324931b3edfeb0245a862c   \n",
       "\n",
       "                                                             basename  \\\n",
       "0            unconditioned_5_L001-ds.46edbd96addd4c7bbcf669fbdd201e29   \n",
       "1        conditioned_DMSO_22_L002-ds.3f381fa67ef345fe81257f9d7fc3ed00   \n",
       "2        conditioned_DMSO_21_L002-ds.d6b480e7b7104216b998a896f63665dd   \n",
       "3  conditioned_Belinostat_18_L004-ds.898be0a8019544ce8e9206a811da1756   \n",
       "4        conditioned_DMSO_19_L004-ds.f14158dcc3324931b3edfeb0245a862c   \n",
       "\n",
       "                                             fastq_R1  \\\n",
       "0             unconditioned-5_S5_L001_R1_001.fastq.gz   \n",
       "1        conditioned-DMSO-22_S16_L002_R1_001.fastq.gz   \n",
       "2        conditioned-DMSO-21_S15_L002_R1_001.fastq.gz   \n",
       "3  conditioned-Belinostat-18_S12_L004_R1_001.fastq.gz   \n",
       "4        conditioned-DMSO-19_S13_L004_R1_001.fastq.gz   \n",
       "\n",
       "                                             fastq_R2               condition  \\\n",
       "0             unconditioned-5_S5_L001_R2_001.fastq.gz           unconditioned   \n",
       "1        conditioned-DMSO-22_S16_L002_R2_001.fastq.gz        conditioned_DMSO   \n",
       "2        conditioned-DMSO-21_S15_L002_R2_001.fastq.gz        conditioned_DMSO   \n",
       "3  conditioned-Belinostat-18_S12_L004_R2_001.fastq.gz  conditioned_Belinostat   \n",
       "4        conditioned-DMSO-19_S13_L004_R2_001.fastq.gz        conditioned_DMSO   \n",
       "\n",
       "  sample_id  lane  \n",
       "0         5  L001  \n",
       "1        22  L002  \n",
       "2        21  L002  \n",
       "3        18  L004  \n",
       "4        19  L004  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load library directories\n",
    "libs = pd.DataFrame(glob(data_dir + \"/*\"), columns=[\"path\"])\n",
    "\n",
    "# Get directory names, which are parsed for sample information\n",
    "libs['basename'] = libs.path.apply(os.path.basename)\n",
    "\n",
    "# Get filename lists of paired-end R1 and R2 fastq files\n",
    "libs['fastq_R1'] = [os.path.basename(glob(path + \"/*_R1_*.fastq*\")[0]) for path in libs.path]\n",
    "libs['fastq_R2'] = [os.path.basename(glob(path + \"/*_R2_*.fastq*\")[0]) for path in libs.path]\n",
    "\n",
    "# name extracted columns from regex capture groups\n",
    "libs[['condition', 'sample_id', 'lane']] = libs.basename.str.extract(r'(.*)_([0-9]+)_(L[0-9]*)')\n",
    "\n",
    "libs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "sound-privilege",
   "metadata": {},
   "outputs": [],
   "source": [
    "# shared arguments to pipeline\n",
    "# for additional arguments and documentation: https://github.com/ENCODE-DCC/atac-seq-pipeline/blob/master/docs/input.md\n",
    "fixed_config = {\n",
    "    'atac.pipeline_type': 'atac',\n",
    "    'atac.genome_tsv': 'https://storage.googleapis.com/encode-pipeline-genome-data/genome_tsv/v4/hg38.tsv',\n",
    "    'atac.paired_end': 'true',\n",
    "    'atac.auto_detect_adapter': 'true',\n",
    "    'atac.enable_xcor': 'true'\n",
    "}\n",
    "\n",
    "# Loop over each condition, generate a config file to submit to ENCODE ATAC-seq pipeline\n",
    "for cond in set(libs.condition):\n",
    "    # subset of biological replicates and lanes associated with treatment condition\n",
    "    libs_sub = libs[libs.condition == cond]\n",
    "    \n",
    "    config = {**fixed_config, 'atac.title': cond}  # ** dictionary unapcking to new dict\n",
    "    \n",
    "    # Loop over biological replicates, parsing fastq files into technical and biological replicates\n",
    "    for k, sample in enumerate(sorted(set(libs_sub.sample_id))):\n",
    "        replicate = libs_sub[libs_sub.sample_id == sample]\n",
    "        \n",
    "        # format list of paths of fastq files for replicate, paired-end\n",
    "        config[f'atac.fastqs_rep{k+1}_R1'] = (replicate.path + \"/\" + replicate.fastq_R1).to_list()\n",
    "        config[f'atac.fastqs_rep{k+1}_R2'] = (replicate.path + \"/\" + replicate.fastq_R2).to_list()\n",
    "\n",
    "    with open(f'config/{cond}.json', 'w', encoding='utf-8') as f:\n",
    "        json.dump(config, f, ensure_ascii=False, indent=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "synthetic-rescue",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
